presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestingPostgreSqlServer.java

```diff
@@ -58,12 +58,7 @@ private static void execute(String url, String sql)
 
     public String getJdbcUrl()
     {
-        return getJdbcUrl(this.dockerContainer::getHostPort);
-    }
-
-    private String getJdbcUrl(DockerContainer.HostPortProvider dockerContainer)
-    {
-        return format("jdbc:postgresql://localhost:%s/%s?user=%s&password=%s", dockerContainer.getHostPort(POSTGRESQL_PORT), DATABASE, USER, PASSWORD);
+        return format("jdbc:postgresql://localhost:%s/%s?user=%s&password=%s", dockerContainer.getMappedPort(POSTGRESQL_PORT), DATABASE, USER, PASSWORD);
```

Comment:
please never hardcode "localhost" because sometimes it is not. `container.getContainerIpAddress()` is a better option.

Note that `PostgreSQLContainer` has `getJdbcUrl()` too

Reply:
Thanks. But it does not allow to embed the user and password in it. I don't want to refactor existings tests to pass user and passwords.

Reply:
It does, via `user`/`password` query params (although it could use more docs 😅):
https://github.com/testcontainers/testcontainers-java/blob/650e5b0a854d400982d0df1a187d4bbec4850caf/modules/jdbc-test/src/test/java/org/testcontainers/jdbc/JDBCDriverTest.java#L53

> jdbc:tc:postgresql:9.6.8://hostname/databasename?user=someuser&password=somepwd

Reply:
Yes. That is the way I am using now (passing user and password as url params). I would prefer to use: `container.getJdbcUrl(user, password)`

Reply:
that's not *exactly* the way you're using it. Note `tc:` prefix.

We register a special magical JDBC driver that handles `tc:` prefix and starts container when first connection is requested, so that you don't need to use `new PostgreSQLContainer()` at all :)

Reply:
Be aware that the container will be terminated as soon as last connection is closed. If you want to keep the container running (for other tests in the same JVM run, for example), you can use the daemon mode:
https://www.testcontainers.org/modules/databases/#running-container-in-daemon-mode

Reply:
> We register a special magical JDBC driver that handles `tc:` prefix and starts container when first connection is requested

Cool feature. I see where it could be applicable.
(I don't see it applicable here.)

Reply:
> you can use the daemon mode:

How can I then trigger container to stop and how to trigger its removal.

> We register a special magical JDBC driver that handles tc: 

I don't like magic. There are only couple of lines and it is very clear when container is started and when it is closed. So I would prefer to leave it as it is. Notice that we created a data source for Presto and tests do not control how Presto will access that data source. Situation when Presto would close "the last" connection and data source would vanish just before the next test is very undesired.